kafka-client/src/main/java/dev/responsive/kafka/internal/db/CassandraKeyValueTable.java

```diff
@@ -338,69 +391,116 @@ public CassandraKVFlushManager init(
+      } else if (result.isEmpty()) {
+        return null;
+      } else {
+        return getValueFromRow(result.get(0));
+      }
     } else {
-      final ByteBuffer value = result.get(0).getByteBuffer(DATA_VALUE.column());
-      return Objects.requireNonNull(value).array();
+      final BoundStatement getQuery = getWithTimestamp
+          .bind()
+          .setByteBuffer(DATA_KEY.bind(), ByteBuffer.wrap(key.get()))
+          .setInt(PARTITION_KEY.bind(), tablePartition)
+          .setLong(TIMESTAMP.bind(), minValidTs);
+      final List<Row> result = client.execute(getQuery).all();
+
+      if (result.size() > 1) {
+        throw new IllegalStateException("Received multiple results for the same key");
+      } else if (result.isEmpty()) {
+        return null;
+      }
+
+      final Row rowResult = result.get(0);
+      final byte[] value = getValueFromRow(rowResult);
+      final TtlDuration ttl = ttlResolver.get().resolveTtl(key, value);
+
+      if (ttl.isFinite()) {
+        final long minValidTsFromValue = streamTimeMs - ttl.toMillis();
+        final long recordTs = rowResult.getLong(TIMESTAMP.column());
+        if (recordTs < minValidTsFromValue) {
+          return null;
+        }
+      }
+
+      return value;
     }
   }
 
+  private byte[] getValueFromRow(final Row row) {
+    return Objects.requireNonNull(row.getByteBuffer(DATA_VALUE.column())).array();
+  }
+
   @Override
   public KeyValueIterator<Bytes, byte[]> range(
       final int kafkaPartition,
       final Bytes from,
       final Bytes to,
-      final long minValidTs
+      final long streamTimeMs
   ) {
     // TODO: explore more efficient ways to serve bounded range queries, for now we have to
     //  iterate over all subpartitions and merge the results since we don't know which subpartitions
     //  hold keys within the given range
     //  One option would be to configure the partitioner with an alternative hasher that's optimized
     //  for range queries with a Comparator-aware key-->subpartition mapping strategy.
+
+    // TODO(sophie): filter by minValidTs if based on key or default only
+    final long minValidTs = 0L;
     final List<KeyValueIterator<Bytes, byte[]>> resultsPerPartition = new LinkedList<>();
     for (final int partition : partitioner.allTablePartitions(kafkaPartition)) {
       final BoundStatement range = this.range
           .bind()
           .setInt(PARTITION_KEY.bind(), partition)
           .setByteBuffer(FROM_BIND, ByteBuffer.wrap(from.get()))
           .setByteBuffer(TO_BIND, ByteBuffer.wrap(to.get()))
-          .setInstant(TIMESTAMP.bind(), Instant.ofEpochMilli(minValidTs));
+          .setLong(TIMESTAMP.bind(), minValidTs);
 
       final ResultSet result = client.execute(range);
       resultsPerPartition.add(Iterators.kv(result.iterator(), CassandraKeyValueTable::rows));
     }
+    // TODO(sophie): filter by minValidTs if based on value
```

Comment:
we should make sure to address these TODOs before cutting a release since this will result in wrong data, I know it's already a big PR so just make sure we're tracking it somewhere